### PR TITLE
Material container & related stuff ui refactors & clean-up

### LINKED ISSUE #72096: Deconstructing ancient protolathe (derelict drone station) destroys materials?
When playing as a drone I wanted to upgrade the lathe before printing more, but deconstructing it seems to just vanish the materials it contains? It wasn't linked to an ore silo yet.

## Round ID:
Hmm sadly can't find the exact ID now

## Reproduction:

Go KS13, build ancient protolathe, stick some materials in, destroy it (before linking it to silo), materials don't pop back out.

### LINKED ISSUE #76151: You can't print things if you have just enough materials
## Reproduction:

Found this after materials rebalance, not sure if this only bs crystals.
Screenshot_165

### LINKED ISSUE #76253: all materials are greyed out in the exosuit fab
## Reproduction: caused by the material refractor stuff 
Screenshot_674

### LINKED ISSUE #76506: RCD silo link consumes excessive material
## Reproduction:

Build, for example, a 10x6 space room using a silo-linked RCD, with the silo containing around 1300 sheets of metal. Not that even just putting down the floors and walling it off will take more resources than the silo has.

I suspect something wasn't changed with the PR that reduced sheets from 2000 material to 100 material, and the RCD is still consuming as if 1 sheet = 2000 silo material.

### PR DESCRIPTION
**Pull request by SyncIt21**
## About The Pull Request

**1. Material container clean-up & refactor**
- Replaced `total_amount` var with `total_amount()` proc, this var can be easily computed by summing up all material amounts rather than storing it as a var which is tedious to update & keep track of when materials are added/removed
- Removed unused  procs `transer_amt_to()`, `can_insert_amount_mat()`, and `get_categories()`. These procs are not used anywhere in the codebase so let's remove them & make some space. 
- Callbacks are replaced with signals, the callbacks don't need to be explicitly garbage collected & having macros & procs marked with `SIGNAL_HANDLER` makes your intentions more readable & explicit.
- Fixes #76151 
All material adding, removal, checking operations are "Integer" operations, i.e. the final value is rounded & them made 1 if the final value is 0 using the macro `OPTIMIZE_COST`[coudn't come up with a better name]. No more dealing with decimal value materials
The problem was after the protolathe was upgraded with better parts all the design costs were multiplied with a decimal `efficiency_coeff` value, this means even though in the UI the cost was displayed as 60 bluespace crystals its actual cost was `60.0001` something in the backend causing this check for materials to fail & print the error message.
- Replaced `GetComponent(/datum/component_material_container)` with just a simple ref to the material container when adding the component, so we can save some overhead from calling this proc
- Gave all procs a ton of documentation with documentation having documentation
- Fixes #76506 RCD and other devices that uses the silo link upgrade now have the correct material usages
- Fixes #72096. It wasn't just a problem with ancient protolathe but with all machines that used `datum/component/remote_materials`  the problem was remote materials would add an instance of   `datum/remote/material_container` if it wanted to use local storage but this component would get added before `datum/component/remote_materials` could be registered i.e. it comes before remote_materials in the component list. So when the machine is destroyed it will first destroy `material_container` & then `remote_materials` therefore destroying the materials before they could get ejected
- Silo link is established when parent is registered with remote materials raher than adding an external timer which is faster
- Everything that uses a material container will auto eject their sheets when destroyed
- Moved this & remote materials into its own folder for better organization

**2. Material UI Changes**
- Removed the x25 & x50 print buttons from the autolathe, now they just have x5 & x10 buttons like the protolathe, These buttons were of no use since you could just type the exact amount you want to print in the `[Max: <some amount>]` side bar. The code to compute these buttons was just plain right nasty & some of it unused in the UI. 
- The material eject button in the material bar does not gray out when you can eject exactly one sheet
- All material cost are integer values rounded
- Fixes #76253 Exosuit Fabricator sends the material container static data to the UI so its material bar is not greyed out when there are sufficient materials to eject
- Component printer material bar sends the material container static data to the UI so its material bar is not greyed out when there are sufficient materials to eject
- Autolathe Material bars now display number of sheets available
- Max printable amount of items are now computed & updated correctly in the UI. They were displaying wrong values & now get updated when items are printed, materials are removed
- Silo hold actually works now. When a machine is put on hold it calls this proc
https://github.com/tgstation/tgstation/blob/e929cf39cded5207d63df1fa8521f41f2816b383/code/datums/components/remote_materials.dm#L78-L87
Notice how the key is `src` so we should be consistent during checking if a machine is on hold using the same `src` var. But for some reason we did dumb shit like this
https://github.com/tgstation/tgstation/blob/e929cf39cded5207d63df1fa8521f41f2816b383/code/datums/components/remote_materials.dm#L150-L153
What is category? Why do we care for the area the machine is in? None of it made sense so i removed all that junk and just made it check for `src` like it should
- Removed redundant `removable` & `sheets` var from the material container ui_data. These vars are unused in the UI
-  If an item does not have the required materials then upon clicking that item you will not get any error message but instead nothing happens

## Changelog
:cl:
fix: items can be printed from autolathe & protolathe when the exact material amounts are present in them after upgrading
fix: max printable amount now shows the correct value & updates when items are printed, materials are removed in the autolathe & protolathe 
fix: component printer material bar is not greyed out when there are sufficient materials to eject
fix: rcd and other devices that uses the silo link upgrade now have the correct material usages
fix: silo hold actually works
fix: machines using local storage to hold materials will eject it's materials as sheets when deconstructed/destroyed
refactor: Autolathe Material bars now display number of sheets available
refactor: printing an item that does not have enough materials will fail silently with no error messages
refactor: Drone dispenser will eject sheets upon deconstruction
refactor: all things that store materials will auto ejects its sheets(if there is sufficient material) when destroyed
refactor: inserting an item into the material container will display the units consumed as sheets not absolute units
refactor: removed x25 & x50 print buttons from the autolathe

### CODE REVIEW

**Comment by SyncIt21** [comment]
Messed up a merge commit, Be right back

**Comment by SyncIt21** [comment]
Actually

**Comment by Sealed101** [comment]
take a look at #76253 while you're here please

**Comment by SyncIt21** [comment]
> take a look at #76253 while you're here please

That bug report while still a bug is misleading. It wasn't broken by #76020 but rather not covered by it. Lathes function as expected it's the "Exosuit Fabricator" specifically that is broken. I'll fix it anyway

**Comment by SyncIt21** [comment]
Fixed

**Comment by SyncIt21** [comment]
This has been sitting for a while now, @ArcaneMusic since this deals with materials and stuff maybe your intrested in a review?

File: code/datums/components/material/material_container.dm [hunk 1]
```diff
@@ -119,27 +107,137 @@
...
+ * - [weapon][obj/item]: the item you are trying to insert
+ * - multiplier: The multiplier for the materials being inserted
+ * - breakdown_flags: The breakdown bitflags that will be used to retrieve the materials from the source
+ */
+/datum/component/material_container/proc/insert_item(obj/item/weapon, multiplier = 1, breakdown_flags = mat_container_flags)
+	if(QDELETED(weapon))
+		return MATERIAL_INSERT_ITEM_NO_MATS
+	multiplier = CEILING(multiplier, 0.01)
+
+	var/obj/item/target = weapon
```
Comment (ArcaneMusic) [inline code/datums/components/material/material_container.dm:196]: You're not typecasting target or weapon to anything different, so you could reasonably use weapon for all target instances or rename weapon to target, either or.

**Review by ArcaneMusic — CHANGES_REQUESTED** [review]
Starting to get through this, just got back and this is a big PR so I'll try and hit more of this after, anyone extra wants to help be my guest

File: code/datums/components/material/material_container.dm (hunk 1 above)
Comment (SyncIt21) [reply code/datums/components/material/material_container.dm:196]: This isn't for typecasting reasons. target holds the item we are trying to insert and it differs from weapon when the stack is split if there isn't enough space
https://github.com/tgstation/tgstation/blob/b3f22e0a7b90ac12217a9382ea2308a6bb1314cf/code/datums/components/material/material_container.dm#L211
Didn't want to confuse it with weapon for clarity sake

File: code/game/machinery/autolathe.dm [hunk 2]
```diff
@@ -153,84 +136,81 @@
...
-			to_chat(usr, span_alert("Not enough materials for this operation."))
+		if(!materials.has_materials(materials_used, coeff, multiplier))
+			say("Not enough materials for this operation!.")
+			return
+
+		//use power
+		var/total_amount = 0
+		for(var/material in being_built.materials)
+			total_amount += being_built.materials[material]
+		var/power = max(active_power_usage, (total_amount) * multiplier / 5) // Change this to use all materials
```
Comment (ArcaneMusic) [inline code/game/machinery/autolathe.dm:198]: Does this still need to be changed?

**Review by ArcaneMusic — CHANGES_REQUESTED** [review]
Just went through everything but the TGUI, found the one thing but aside from that seems fairly clean. I'll ask for a quick TM just to make sure we don't break prod or anything like that (again) (I have a habit of that sometimes)

File: code/game/machinery/autolathe.dm (hunk 2 above)
Comment (SyncIt21) [reply code/game/machinery/autolathe.dm:198]: That comment is outdated so no nothing needs to be changed, i copied & pasted it by accident I'll remove it

**Review by ArcaneMusic — APPROVED** [review]
Observed a few TM rounds, no issues observed or gleaned by a 60+ player server. So LGTM

**Pull request merged**